File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        vis = set()
        def dfs(node):
            if node in vis:
                return 0
            vis.add(node)
            temp = 0
            for x in adj[node]:
                temp += dfs(x)
            if temp > 0:
                return temp + 2
            return 2 if hasApple[node] else 0

        return max(dfs(0) - 2, 0)

File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.cpp
class Solution {
public:
    vector<vector<int>> adj;
    int dfs(vector<bool>& hasApple,int node,int d,int prev){
        int ans=0,temp;
        for(auto i:adj[node])
	    if(i!=prev){
	        temp=dfs(hasApple,i,d+1,node);
	        if(temp)
		    ans+=temp-d;
	    }
        return ans||hasApple[node]?ans+d:0;

    }
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple){
        adj.resize(n);
        for(vector<int> &e:edges)
            adj[e[0]].push_back(e[1]),adj[e[1]].push_back(e[0]);
        return dfs(hasApple,0,0,-1)*2;     
    }
};
